File: packages/core/skills/extracting-requirements/scripts/aggregate_stories.py
#!/usr/bin/env python3
"""Aggregate extracted story card JSONs into per-epic requirement files.

Usage: aggregate_stories.py -o <output-dir> <json-file>...

Takes one or more JSON files (each a list of story objects or {"stories": [...]}),
deduplicates by title, groups into epics by epic_theme, assigns stable IDs
(STORY-NNNN, EPIC-NNN), and writes one file per epic to the output directory.
"""
import argparse
import json
import sys
from collections import OrderedDict
from pathlib import Path


def load_stories(paths: list[Path]) -> list[dict]:
    """Load and combine stories from multiple JSON files."""
    all_stories: list[dict] = []
    for p in paths:
        data = json.loads(p.read_text())
        if isinstance(data, list):
            all_stories.extend(data)
        elif isinstance(data, dict) and "stories" in data:
            all_stories.extend(data["stories"])
        else:
            print(f"warning: {p} has unexpected format, skipping", file=sys.stderr)
    return all_stories


def _story_body_key(story: dict):
    """The fields that must agree for two same-titled stories to be the same
    requirement rather than a title collision between different ones."""
    return (
        story.get("i_want", "").strip(),
        story.get("so_that", "").strip(),
        json.dumps(story.get("acceptance_criteria", []), sort_keys=True),
    )


def dedup_stories(stories: list[dict]) -> list[dict]:
    """Deduplicate stories that agree on epic_theme, title, and body
    (i_want/so_that/acceptance_criteria). Merges sources from duplicates.

    Title alone is not a safe key: two different requirements can get the
    same short title from an extraction subagent, and merging them would
    silently drop one epic's requirement and misattribute its citations to
    the survivor. An empty title is never treated as a match, even against
    another empty title — format_epic_file's 'Untitled' fallback is a
    display default, not a shared identity.
    """
    seen: "OrderedDict[tuple, dict]" = OrderedDict()
    empty_title_count = 0
    for story in stories:
        title = story.get("title", "").strip()
        if not title:
            empty_title_count += 1
            key = ("__no_title__", empty_title_count)
        else:
            theme = story.get("epic_theme", "Uncategorized").strip()
            key = (theme, title, _story_body_key(story))
        if key in seen:
            existing_sources = seen[key].get("sources", [])
            for src in story.get("sources", []):
                if src not in existing_sources:
                    existing_sources.append(src)
            seen[key]["sources"] = existing_sources
        else:
            seen[key] = dict(story)  # copy to avoid mutating input
    return list(seen.values())


def group_into_epics(stories: list[dict]) -> dict[str, list[dict]]:
    """Group stories by epic_theme. Returns ordered dict {theme: [stories]}."""
    epics: dict[str, list[dict]] = OrderedDict()
    for story in stories:
        theme = story.get("epic_theme", "Uncategorized").strip()
        if theme not in epics:
            epics[theme] = []
        epics[theme].append(story)
    return epics


def assign_ids(epics: dict[str, list[dict]]) -> None:
    """Assign stable EPIC-NNN and STORY-NNNN IDs to all stories in place."""
    story_counter = 1
    epic_counter = 1
    for theme, stories in epics.items():
        epic_id = f"EPIC-{epic_counter:03d}"
        epic_counter += 1
        for story in stories:
            story["_id"] = f"STORY-{story_counter:04d}"
            story["_epic_id"] = epic_id
            story["_epic_theme"] = theme
            story_counter += 1


def format_epic_file(epic_id: str, theme: str, stories: list[dict]) -> str:
    """Format one epic as a standalone markdown file."""
    lines: list[str] = []

    story_ids = [s["_id"] for s in stories]

    primary_sources: set[str] = set()
    for s in stories:
        for src in s.get("sources", []):
            if isinstance(src, dict):
                primary_sources.add(src.get("file", ""))
            elif isinstance(src, str):
                primary_sources.add(src)

    lines.append(f"# {epic_id} — {theme}")
    lines.append("")
    lines.append(f"**Summary:** {theme}")
    lines.append(f"**Stories:** {', '.join(story_ids)}")
    if primary_sources:
        sources_str = ", ".join(f"`{s}`" for s in sorted(primary_sources) if s)
        lines.append(f"**Primary sources:** {sources_str}")
    lines.append(f"**Status:** 0/{len(stories)} done")
    lines.append("")

    for story in stories:
        lines.append(f"## {story['_id']}")
        lines.append("")
        lines.append(f"**Epic:** {story['_epic_id']} — {story['_epic_theme']}")
        lines.append(f"**Title:** {story.get('title', 'Untitled')}")
        lines.append("")
        lines.append(f"**As a** {story.get('as_a', 'user')}")
        lines.append(f"**I want** {story.get('i_want', 'this feature')}")
        lines.append(f"**So that** {story.get('so_that', 'I can benefit')}")
        lines.append("")
        lines.append("**Acceptance criteria:**")
        for ac in story.get("acceptance_criteria", []):
            if isinstance(ac, dict):
                ac_text = ac.get("text", "")
                ac_id = ac.get("id", "")
                impact = ac.get("behavioral_impact", "")
                seam = ac.get("proof_seam", "")
                line = f"- {ac_id}: {ac_text}"
                if impact:
                    line += f" · impact:`{impact}`"
                if seam:
                    line += f" · seam:`{seam}`"
                lines.append(line)
            else:
                # Legacy plain-string AC format
                lines.append(f"- {ac}")
        lines.append("")
        lines.append("**Sources:**")
        for src in story.get("sources", []):
            if isinstance(src, dict):
                file_ref = src.get("file", "unknown")
                line_ref = src.get("lines", "")
                ref = f"`{file_ref}:{line_ref}`" if line_ref else f"`{file_ref}`"
                lines.append(f"- {ref}")
            elif isinstance(src, str):
                lines.append(f"- `{src}`")
        lines.append("")
        lines.append("**Status:** pending")
        lines.append("")

    return "\n".join(lines)


def main() -> int:
    parser = argparse.ArgumentParser(description="Aggregate stories into per-epic files")
    parser.add_argument("-o", "--output-dir", required=True,
                        help="Directory to write per-epic files (created if needed)")
    parser.add_argument("json_files", nargs="+", help="Extracted story JSON files")
    args = parser.parse_args()

    paths = [Path(p) for p in args.json_files]
    for p in paths:
        if not p.exists():
            print(f"error: file not found: {p}", file=sys.stderr)
            return 2

    stories = load_stories(paths)
    if not stories:
        print("error: no stories found in input files", file=sys.stderr)
        return 1

    deduped = dedup_stories(stories)
    epics = group_into_epics(deduped)
    assign_ids(epics)

    out_dir = Path(args.output_dir)
    out_dir.mkdir(parents=True, exist_ok=True)

    for theme, epic_stories in epics.items():
        epic_id = epic_stories[0]["_epic_id"]
        content = format_epic_file(epic_id, theme, epic_stories)
        out_path = out_dir / f"{epic_id}.md"
        out_path.write_text(content)
        print(f"wrote {out_path} ({len(epic_stories)} stories)")

    total = sum(len(s) for s in epics.values())
    print(f"OK: {len(epics)} epics, {total} stories")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: packages/core/skills/extracting-requirements/scripts/aggregate_stories.mjs
import {
  existsSync,
  mkdirSync,
  readFileSync,
  readdirSync,
  unlinkSync,
  writeFileSync,
} from "node:fs";
import { basename, join, resolve } from "node:path";
import { pathToFileURL } from "node:url";

function stableJson(value) {
  if (Array.isArray(value)) return `[${value.map(stableJson).join(",")}]`;
  if (value && typeof value === "object") {
    return `{${Object.keys(value)
      .sort()
      .map((key) => `${JSON.stringify(key)}:${stableJson(value[key])}`)
      .join(",")}}`;
  }
  return JSON.stringify(value);
}

export function loadStories(paths, warn = (message) => process.stderr.write(`${message}\n`)) {
  const stories = [];
  for (const path of paths) {
    const data = JSON.parse(readFileSync(path, "utf8"));
    if (Array.isArray(data)) stories.push(...data);
    else if (data && typeof data === "object" && "stories" in data) stories.push(...data.stories);
    else warn(`warning: ${path} has unexpected format, skipping`);
  }
  return stories;
}

export function storyBodyKey(story) {
  return [
    (story.i_want ?? "").trim(),
    (story.so_that ?? "").trim(),
    stableJson(story.acceptance_criteria ?? []),
  ];
}

export function dedupStories(stories) {
  const seen = new Map();
  let blankTitleCount = 0;

  for (const story of stories) {
    const title = (story.title ?? "").trim();
    const key = title
      ? stableJson([(story.epic_theme ?? "Uncategorized").trim(), title, storyBodyKey(story)])
      : stableJson(["__no_title__", ++blankTitleCount]);

    const existing = seen.get(key);
    if (existing) {
      const sourceKeys = new Set(existing.sources.map(stableJson));
      for (const source of story.sources ?? []) {
        const sourceKey = stableJson(source);
        if (!sourceKeys.has(sourceKey)) {
          existing.sources.push(source);
          sourceKeys.add(sourceKey);
        }
      }
    } else {
      seen.set(key, { ...story, sources: [...(story.sources ?? [])] });
    }
  }

  return [...seen.values()];
}

export function groupIntoEpics(stories) {
  const epics = new Map();
  for (const story of stories) {
    const theme = (story.epic_theme ?? "Uncategorized").trim();
    const grouped = epics.get(theme);
    if (grouped) grouped.push(story);
    else epics.set(theme, [story]);
  }
  return epics;
}

export function assignIds(epics) {
  let storyCounter = 1;
  let epicCounter = 1;
  for (const [theme, stories] of epics) {
    const epicId = `EPIC-${String(epicCounter).padStart(3, "0")}`;
    epicCounter += 1;
    for (const story of stories) {
      story._id = `STORY-${String(storyCounter).padStart(4, "0")}`;
      story._epic_id = epicId;
      story._epic_theme = theme;
      storyCounter += 1;
    }
  }
  return epics;
}

export function formatEpicFile(epicId, theme, stories) {
  const lines = [];
  const primarySources = new Set();
  for (const story of stories) {
    for (const source of story.sources ?? []) {
      primarySources.add(typeof source === "object" && source !== null ? (source.file ?? "") : source);
    }
  }

  lines.push(`# ${epicId} — ${theme}`);
  lines.push("");
  lines.push(`**Summary:** ${theme}`);
  lines.push(`**Stories:** ${stories.map((story) => story._id).join(", ")}`);
  if (primarySources.size > 0) {
    const sources = [...primarySources]
      .filter(Boolean)
      .sort()
      .map((source) => `\`${source}\``)
      .join(", ");
    lines.push(`**Primary sources:** ${sources}`);
  }
  lines.push(`**Status:** 0/${stories.length} done`);
  lines.push("");

  for (const story of stories) {
    lines.push(`## ${story._id}`);
    lines.push("");
    lines.push(`**Epic:** ${story._epic_id} — ${story._epic_theme}`);
    lines.push(`**Title:** ${story.title ?? "Untitled"}`);
    lines.push("");
    lines.push(`**As a** ${story.as_a ?? "user"}`);
    lines.push(`**I want** ${story.i_want ?? "this feature"}`);
    lines.push(`**So that** ${story.so_that ?? "I can benefit"}`);
    lines.push("");
    lines.push("**Acceptance criteria:**");
    for (const criterion of story.acceptance_criteria ?? []) {
      if (criterion && typeof criterion === "object") {
        let line = `- ${criterion.id ?? ""}: ${criterion.text ?? ""}`;
        if (criterion.behavioral_impact) line += ` · impact:\`${criterion.behavioral_impact}\``;
        if (criterion.proof_seam) line += ` · seam:\`${criterion.proof_seam}\``;
        lines.push(line);
      } else {
        lines.push(`- ${criterion}`);
      }
    }
    lines.push("");
    lines.push("**Sources:**");
    for (const source of story.sources ?? []) {
      if (source && typeof source === "object") {
        const file = source.file ?? "unknown";
        const reference = source.lines ? `\`${file}:${source.lines}\`` : `\`${file}\``;
        lines.push(`- ${reference}`);
      } else if (typeof source === "string") {
        lines.push(`- \`${source}\``);
      }
    }
    lines.push("");
    lines.push("**Status:** pending");
    lines.push("");
  }

  return lines.join("\n");
}

function usage(program) {
  return `usage: ${program} [-h] -o OUTPUT_DIR json_files [json_files ...]\n`;
}

function help(program) {
  return [
    usage(program).trimEnd(),
    "",
    "Aggregate stories into per-epic files",
    "",
    "positional arguments:",
    "  json_files            Extracted story JSON files",
    "",
    "options:",
    "  -h, --help            show this help message and exit",
    "  -o, --output-dir OUTPUT_DIR",
    "                        Directory to write per-epic files (created if needed)",
    "",
  ].join("\n");
}

function parseArgs(args) {
  let outputDir;
  const jsonFiles = [];
  const unrecognized = [];

  for (let index = 0; index < args.length; index += 1) {
    const arg = args[index];
    if (arg === "-h" || arg === "--help") return { help: true };
    if (arg === "-o" || arg === "--output-dir") {
      const value = args[++index];
      if (value === undefined || value.startsWith("-")) {
        return { error: "argument -o/--output-dir: expected one argument" };
      }
      outputDir = value;
    } else if (arg.startsWith("--output-dir=")) {
      outputDir = arg.slice("--output-dir=".length);
    } else if (arg.startsWith("-o") && arg.length > 2) {
      outputDir = arg.slice(2);
    } else if (arg.startsWith("-")) {
      unrecognized.push(arg);
    } else {
      jsonFiles.push(arg);
    }
  }

  if (unrecognized.length > 0) return { error: `unrecognized arguments: ${unrecognized.join(" ")}` };
  const required = [];
  if (!outputDir) required.push("-o/--output-dir");
  if (jsonFiles.length === 0) required.push("json_files");
  if (required.length > 0) {
    return { error: `the following arguments are required: ${required.join(", ")}` };
  }
  return { outputDir, jsonFiles };
}

export async function main(args) {
  const program = basename(process.argv[1] ?? "aggregate_stories.mjs");
  const options = parseArgs(args);
  if (options.help) {
    process.stdout.write(help(program));
    return 0;
  }
  if (options.error) {
    process.stderr.write(`${usage(program)}${program}: error: ${options.error}\n`);
    return 2;
  }

  for (const path of options.jsonFiles) {
    if (!existsSync(path)) {
      process.stderr.write(`error: file not found: ${path}\n`);
      return 2;
    }
  }

  const stories = loadStories(options.jsonFiles);
  if (stories.length === 0) {
    process.stderr.write("error: no stories found in input files\n");
    return 1;
  }

  const epics = assignIds(groupIntoEpics(dedupStories(stories)));
  mkdirSync(options.outputDir, { recursive: true });
  for (const name of readdirSync(options.outputDir)) {
    if (/^EPIC-.*\.md$/.test(name)) unlinkSync(join(options.outputDir, name));
  }

  for (const [theme, epicStories] of epics) {
    const epicId = epicStories[0]._epic_id;
    const outputPath = join(options.outputDir, `${epicId}.md`);
    writeFileSync(outputPath, formatEpicFile(epicId, theme, epicStories));
    process.stdout.write(`wrote ${outputPath} (${epicStories.length} stories)\n`);
  }
  const total = [...epics.values()].reduce((count, storiesInEpic) => count + storiesInEpic.length, 0);
  process.stdout.write(`OK: ${epics.size} epics, ${total} stories\n`);
  return 0;
}

const isDirect = process.argv[1] && pathToFileURL(resolve(process.argv[1])).href === import.meta.url;
if (isDirect) process.exitCode = await main(process.argv.slice(2));
